sound/soc/sof/loader.c

```diff
@@ -495,6 +597,26 @@ int snd_sof_load_firmware_raw(struct snd_sof_dev *sdev)
 			fw_filename);
 	}
 
+	/* check for extended manifest */
+	ext_man_size = snd_sof_ext_man_size(plat_data->fw);
+	if (ext_man_size > 0) {
+		ret = snd_sof_fw_ext_man_parse(sdev, plat_data->fw);
+
+		/* when no error occurred, drop extended manifest */
+		if (!ret)
+			plat_data->fw_offset = ext_man_size;
+		else
+			dev_err(sdev->dev, "error: firmware %s contains unsupported or invalid extended manifest: %d\n",
+				fw_filename, ret);
+	} else if (!ext_man_size) {
+		/* No extended manifest, so nothing to skip during FW load */
+		dev_dbg(sdev->dev, "firmware doesn't contain extended manifest\n");
+	} else {
+		ret = ext_man_size;
+		dev_err(sdev->dev, "error: firmware %s contains unsupported or invalid extended manifest: %d\n",
+			fw_filename, ret);
+	}
+
```

Comment:
@ktrzcinx With the risk of repeating an old disucssion again, can you refresh why we have the same check on snd_sof_ext_man_size() two times? Do we end up calling snd_sof_fw_ext_man_parse() from other places, or? Because now when I look at this, I can't see how the errors in ext_man_parse() could be hit.

Reply:
In `snd_sof_load_firmware_raw()` there was check from the begin and in `snd_sof_fw_ext_man_parse()` was added because of @plbossart suggestion: https://github.com/thesofproject/linux/pull/1818#discussion_r382863466

Reply:
@ktrzcinx Hmm, I went through #1818 comments and couldn't really find a good reason for the double checks. @plbossart  ?

@ktrzcinx What if you pass "ext_man_size" as an argument to snd_sof_fw_ext_man_parse(). Then it's immediately obvious that is must be >0 and you can just remove L168-L178. If anyone else starts using this funciton later, they have to pass a valid ext_man_size.

Reply:
When ext_man_size will be passed as an argument it can quite easily loose sync with fw content, so this solution is error prone. In my opinion, if double error is a problem then check should be done only in parsing function and after detect lack of ext_man there should be dbg info instead of error.

Reply:
@ktrzcinx Fair enough, so then maybe the best solution is to put the checks within ext_man_parse(). Some defensive programming is ok, but at least it looks a bit odd to have identical checks with verbose error traces first in a private static inline function, which is only called from one place, and then the same checks and trace strings in the caller as well.

Reply:
updated